### PR TITLE
[billing] provide human readable price info

### LINKED ISSUE #15786: Change UBP plan for teams to the same as individuals.
In the context of #15665 we want to make sure that users still have freemium and can trial usage of Gitpod with their team.

This issue is about making two changes:
1) When creating an org for a user because they don't have one we apply 500 free credits per month.
2) Change the stripe price for teams to be the same as the one for individuals (i.e. 9$/per month which includes 1000 credits)

depends on https://github.com/gitpod-io/gitpod/issues/15799

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
Use a human-readable string configured on the stripe price to inform the user of their current subscription.
Aligns prices between teams and individuals, so that teams get the first 1000 credits for 9 USD as well.

Screenshot 2023-01-19 at 17 49 51

## Related Issue(s)

Fixes #15786

Depends on
- https://github.com/gitpod-io/ops/issues/7904

Also related 
 - https://github.com/gitpod-io/ops/pull/7799
 - https://github.com/gitpod-io/ops/pull/7800

## How to test

Check subscription for a new team is $9/month and verify the information presented makes sense before and after subscribing.

## Release Notes

```release-note
Teams get the first 1000 credits for 9 USD/€
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-payment
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx [hunk 1]
```diff
@@ -226,72 +231,43 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
...
-                            </div>
-                        </div>
+                        {usageLimit > 0 && (
+                            <>
+                                <div className="uppercase text-sm text-gray-400 dark:text-gray-500">Current Plan</div>
+                                <div className="mt-1 text-xl font-semibold flex-grow text-gray-600 dark:text-gray-400">
+                                    {usageLimit > 500 ? "Open Source" : "Free"}
+                                </div>
+                                <div className="mt-4 flex space-x-1 text-gray-400 dark:text-gray-500">
+                                    <Check className="m-0.5 w-5 h-5 text-orange-500" />
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:243]: **suggestion**: This was probably changed in a past PR, but could we increase the size of the checkmark here?

```suggestion
                                    <Check className="m-0.5 w-8 h-5 text-orange-500" />
```

| BEFORE | AFTER |
|-|-|
| checkmark-before | checkmark-after |

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx [hunk 2]
```diff
@@ -226,72 +231,43 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
...
-                                <span className="font-bold text-gray-500 dark:text-gray-400">{usageLimit} credits</span>
-                                <span>
-                                    {usageLimit / 10} hours of Standard workspace usage.{" "}
-                                    <a className="gp-link" href="https://www.gitpod.io/pricing#cost-estimator">
-                                        Estimate costs
-                                    </a>
-                                </span>
-                            </div>
-                        </div>
+                        {usageLimit > 0 && (
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:236]: **issue**: Not sure where to comment about this but before upgrading the layout seems broken with missing text information.

| Before upgrading | After upgrading |
|-|-|
| billing-before | billing-after |

File: components/usage/pkg/apiv1/billing.go [hunk 3]
```diff
@@ -382,6 +382,61 @@ func (s *BillingService) CancelSubscription(ctx context.Context, in *v1.CancelSu
...
+	priceID, err := getPriceIdentifier(attributionID, stripeCustomer, s)
+	if err != nil {
+		log.Errorf("Failed to get price identifier for attribution ID %s: %s", attributionId, err.Error())
+		return defaultPriceId
+	}
+	return priceID
+}
+
+func (s *BillingService) GetPriceInformation(ctx context.Context, req *v1.GetPriceInformationRequest) (*v1.GetPriceInformationResponse, error) {
+	priceID := s.getPriceId(ctx, req.GetAttributionId())
```
Comment (Siddhant-K-code) [inline components/usage/pkg/apiv1/billing.go:426]: **Suggestion**: Check if required field is available/not


```suggestion
	if req.GetAttributionId() == "" {
		return nil, status.Error(codes.InvalidArgument, "attribution_id is required")
	}
	priceID := s.getPriceId(ctx, req.GetAttributionId())
```

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 4]
```diff
@@ -2111,6 +2111,28 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
+        } else {
+            if (attrId.userId !== user.id) {
+                throw new ResponseError(
+                    ErrorCodes.PERMISSION_DENIED,
+                    "Cannot create Stripe customer profile for another user",
+                );
+            }
+            await this.ensureStripeApiIsAllowed({ user });
+        }
+        return this.stripeService.getPriceInformation(attributionId);
```
Comment (Siddhant-K-code) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2133]: **Suggestion**: We Can follow the consistent approach (used above) to return err if Price Information is not available. (Would be easy to debug in future)

```suggestion
        const priceInformation = this.stripeService.getPriceInformation(attributionId);
        if (!priceInformation) {
            throw new ResponseError(
                ErrorCodes.NOT_FOUND,
                `No price information found for attributionId '${attributionId}'`,
            );
        }
        return priceInformation;
```

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx [hunk 5]
```diff
@@ -226,72 +231,43 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
...
+                            <>
+                                <div className="uppercase text-sm text-gray-400 dark:text-gray-500">Current Plan</div>
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:238]: **question**: Should we surface anything about the current plan when a user has not upgraded? If we don't want to name the plan like Pay-as-you-go we could probably drop the extra card element on top.

| CURRENT (Before upgrading on production) | AFTER (Before upgrading on this PR) |
|-|-|
| billing-production | billing-current |

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx [hunk 6]
```diff
@@ -312,65 +288,24 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
...
-                                </>
-                            )}
+                            <div className="mt-1 text-xl font-semibold flex-grow text-gray-800 dark:text-gray-100">
+                                Pay-as-you-go
+                            </div>
+                            <div className="mt-4 flex space-x-1 text-gray-400 dark:text-gray-500">
+                                <Check className="m-0.5 w-5 h-5 text-orange-500" />
+                                <div className="flex flex-col">
+                                    <span>
+                                        {priceInformation}{" "}
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:298]: **question**: I know this is coming from elsewhere, but could we drop the two decimals (`$9.00`) on the monthly price? Added also some more typo nitpicks if that's possible.

```
// BEFORE
$9.00/month for the first 1000 credits. Afterward pay-as-you-go for $0.036 per credit.

// AFTER
$9 per month for the first 1,000 credits and pay-as-you-go after that for $0.036 per credit.
```

**Review by gtsiolis — COMMENTED** [review]
Left some UX comments below. 🍊 

ok

**Review by mads-hartmann — APPROVED** [review]
DevX-owned changes LGTM 🚀

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx (hunk 6 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageBasedBillingConfig.tsx:298]: please suggest that [here](https://github.com/gitpod-io/ops/issues/7904)

Comment (gtsiolis) [reply components/dashboard/src/components/UsageBasedBillingConfig.tsx:298]: Copied the comment in https://github.com/gitpod-io/ops/issues/7904#issuecomment-1398313688. Resolving this. 🏓

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 7]
```diff
@@ -2111,6 +2111,28 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
+        }
+        let team: Team | undefined;
+        if (attrId.kind === "team") {
+            team = await this.guardTeamOperation(attrId.teamId, "update");
+            await this.ensureStripeApiIsAllowed({ team });
+        } else {
+            if (attrId.userId !== user.id) {
+                throw new ResponseError(
+                    ErrorCodes.PERMISSION_DENIED,
+                    "Cannot create Stripe customer profile for another user",
```
Comment (geropl) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2128]: That error message looks weird, maybe copy'n'paste?

File: components/usage/pkg/apiv1/billing.go [hunk 8]
```diff
@@ -382,6 +382,61 @@ func (s *BillingService) CancelSubscription(ctx context.Context, in *v1.CancelSu
...
+	stripeCustomer, err := s.stripeClient.GetCustomer(ctx, customer.Customer.Id)
+	if err != nil {
+		log.Errorf("Failed to get customer infromation from stripe for customer ID %s: %s", customer.Customer.Id, err.Error())
+		return defaultPriceId
+	}
+
+	// if the customer has an active subscription, return that information
+	for _, subscription := range stripeCustomer.Subscriptions.Data {
+		if subscription.Status != "canceled" {
+			return subscription.Plan.ID
```
Comment (geropl) [inline components/usage/pkg/apiv1/billing.go:414]: Not familiar with the Stripe API: The `.Plan.ID` is the same as the `priceID` below? :thinking:

File: components/usage/pkg/stripe/stripe.go [hunk 9]
```diff
@@ -251,6 +251,32 @@ func (c *Client) GetCustomer(ctx context.Context, customerID string) (customer *
 	return customer, nil
 }
 
+func (c *Client) GetPriceInformation(ctx context.Context, priceID string) (price *stripe.Price, err error) {
+	now := time.Now()
+	reportStripeRequestStarted("customer_get")
```
Comment (geropl) [inline components/usage/pkg/stripe/stripe.go:256]: Should be: `reportStripeRequestStarted("prices_get")` or similar.

**Comment by geropl** [comment]
@svenefftinge Code LGTM sans minor stuff above, and UI works for the simple cases. Number of diff lines is quite high, but I see those primarily coming from gRPC so fine with me.
Also understood we need to add the metadata in the first place before rolling this out. (https://github.com/gitpod-io/ops/issues/7904)

My biggest question: What is the impact on pricing, e.g. what's the impact of changing the price IDs in this context? :thinking:

File: components/usage/pkg/apiv1/billing.go (hunk 8 above)
Comment (svenefftinge) [reply components/usage/pkg/apiv1/billing.go:414]: yes

**Comment by svenefftinge** [comment]
> Number of diff lines is quite high, but I see those primarily coming from gRPC so fine with me. 👍 

I agree too much indirection and shoveling. But this is what our current architecture looks like so I followed that.

> Also understood we need to add the metadata in the first place before rolling this out. ([gitpod-io/ops#7904](https://github.com/gitpod-io/ops/issues/7904))

Yep

> My biggest question: What is the impact on pricing, e.g. what's the impact of changing the price IDs in this context? 🤔

What kind of impact you are referring to. Business wise?
FWIW there was [this RFC](https://www.notion.so/gitpod/UBP-Change-pricing-for-Teams-0dbb760ec23043649fbc0f864f4b7b4e). Does that answer that question?

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx (hunk 2 above)
Comment (jldec) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:236]: Why is the estimator only showing when usageLimit > 0?

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx [hunk 10]
```diff
@@ -226,72 +231,43 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
...
-                                <Check className="m-0.5 w-5 h-5" />
-                                <div className="flex flex-col">
-                                    <span className="font-bold">Pay-as-you-go after 1,000 credits</span>
                                     <span>
-                                        {currency === "EUR" ? "€" : "$"}0.36 for 10 credits or 1 hour of Standard
-                                        workspace usage, excluding VAT.
+                                        {priceInformation}{" "}
+                                        <a
+                                            className="gp-link"
+                                            href="https://www.gitpod.io/docs/configure/billing/usage-based-billing"
```
Comment (jldec) [inline components/dashboard/src/components/UsageBasedBillingConfig.tsx:267]: This link should be updated to point to https://www.gitpod.io/docs/configure/billing

**Comment by geropl** [comment]
> > My biggest question: What is the impact on pricing, e.g. what's the impact of changing the price IDs in this context? thinking
> 
> What kind of impact you are referring to. Business wise?
> FWIW there was [this RFC](https://www.notion.so/gitpod/UBP-Change-pricing-for-Teams-0dbb760ec23043649fbc0f864f4b7b4e). Does that answer that question?

Exactly; wasn't aware of that doc but that had all context. :heavy_check_mark:

**Comment by jldec** [comment]
Approve, with the caveat that pricing changes in production need to be coordinated with website changes so that the advertised pricing for teams is the same on www.gitpod.io also. 

This assumes that the prod pricing changes will happen through another processs - not part of this PR.

(Also made 2 comments regarding the estimator link, and the docs link)

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx (hunk 10 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageBasedBillingConfig.tsx:267]: 👍 I have moved and replaced this link with the estimator link.

File: components/dashboard/src/components/UsageBasedBillingConfig.tsx (hunk 2 above)
Comment (jldec) [reply components/dashboard/src/components/UsageBasedBillingConfig.tsx:236]: I think the logic should be show estimator when no limit (before decision to upgrade.)

**Review by geropl — APPROVED** [review]
Approving to unblock, but leaving the lock for you to release. :+1:

**Pull request merged**